src/hotspot/share/cds/filemap.cpp

```diff
@@ -1135,13 +1134,29 @@ bool FileMapInfo::init_from_file(int fd) {
     return false;
   }
 
-  if (header()->header_size() != sz) {
-    log_info(cds)("_header_size expected: " SIZE_FORMAT, sz);
-    log_info(cds)("               actual: " SIZE_FORMAT, header()->header_size());
+  if (header()->header_size() != (unsigned int)runtime_header_size) {
+    log_info(cds)("_header_size expected: " SIZE_FORMAT, runtime_header_size);
+    log_info(cds)("               actual: " UINT32_FORMAT, header()->header_size());
     FileMapInfo::fail_continue("The shared archive file has an incorrect header size.");
     return false;
   }
 
+  if (is_static()) {
+    if (header()->base_archive_path_offset() != 0) {
+      log_info(cds)("_base_archive_path_offset = " UINT32_FORMAT, header()->base_archive_path_offset());
+      fail_continue("_base_archive_path_offset should be 0");
+      return false;
+    }
+  } else {
+    if (header()->base_archive_path_offset() != header()->header_size()) {
+      log_info(cds)("_base_archive_path_offset should be equal to _header_size:");
+      log_info(cds)("  _base_archive_path_offset = " UINT32_FORMAT, header()->base_archive_path_offset());
+      log_info(cds)("  _header_size              = " UINT32_FORMAT, header()->header_size());
+      fail_continue("_base_archive_path_offset should be equal to _header_size");
+      return false;
+    }
+  }
+
```

Comment:
The check from 1144-1158 should be done inside `FileMapInfo::check_archive()`. This function  (`FileMapInfo::init_from_file`) is called *after* we have successfully loaded the header of an archive created by the current JDK. However, -- with JDK-8261455, we will need to read the headers of an archive created by a different JDK version, so we need to do the check at the very beginning, when we are examining the magic value, etc.

Also, the base archive's name should be considered as part of the header, so `header()->header_size()` should include the size of the base archive's name as well. Otherwise, we won't be able to read the entire header, including the base archive name, of an archive created by a different JDK version.

The check for header()->base_archive_path_offset() needs to be expanded:
- it must be smaller than header()->header_size()
- it must point to a valid 0-terminated string -- the string cannot run past the end of the header.

Reply:
The base archive name size should be included in CDSFileMapHeaderBase, not the base archive name itself. If the name as a part of header size, it will complicate the code. We want to read the base archive name from the archive file, this should be done by only reading CDSFileMapHeaderBase from the archive file.